### PR TITLE
8321270: Virtual Thread.yield consumes parking permit

### PR DESCRIPTION
**Pull request by AlanBateman**
When a virtual thread continues after Thread.yield it currently consumes thread's parking permit. This is an oversight, the parking permit should only be consumed when continuing after park.

The changes are straight-forward. The internal "RUNNABLE" state is replaced with UNPARKED and YIELDED state, effectively encoding the previous state. So for the most part, it's just replacing the usages of RUNNABLE. The additional states require refactoring tryGetStackTrace, this is the method that is used for Thread::getStackTrace when the virtual thread is unmounted. It is also changed to not not swallow the REE if the reesubmit fails (tryStackTrace has to resubmit as the task for the thread may run, or the thread unparked, while "suspended" and sampling its stack trace). The changes are a subset of larger changes in the loom repo, future PRs will propose to bring in other changes to get main line up to date.

For testing the existing ThreadAPI has new test cases.

Testing: test1-5. This includes the JVMTI tests as it maps the thread states to JVMTI thread states.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321270](https://bugs.openjdk.org/browse/JDK-8321270): Virtual Thread.yield consumes parking permit (**Bug** - P4)

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16953/head:pull/16953` \
`$ git checkout pull/16953`

Update a local copy of the PR: \
`$ git checkout pull/16953` \
`$ git pull https://git.openjdk.org/jdk.git pull/16953/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16953`

View PR using the GUI difftool: \
`$ git pr show -t 16953`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16953.diff">https://git.openjdk.org/jdk/pull/16953.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16953#issuecomment-1839331418)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/VirtualThread.java [hunk 1]
```diff
@@ -96,37 +96,37 @@ final class VirtualThread extends BaseVirtualThread {
...
     private static final int NEW      = 0;
     private static final int STARTED  = 1;
-    private static final int RUNNABLE = 2;     // runnable-unmounted
-    private static final int RUNNING  = 3;     // runnable-mounted
-
-    // untimed parking
-    private static final int PARKING  = 4;
-    private static final int PARKED   = 5;     // unmounted
-    private static final int PINNED   = 6;     // mounted
+    private static final int RUNNING  = 2;     // runnable-unmounted
```
Comment (sspitsyn) [inline src/java.base/share/classes/java/lang/VirtualThread.java:116]: Just want to make sure this change is intentional.
Before the comment was:  `// runnable-mounted`.

File: src/hotspot/share/classfile/javaClasses.cpp [hunk 2]
```diff
@@ -1990,11 +1990,12 @@ JavaThreadStatus java_lang_VirtualThread::map_state_to_thread_status(int state)
       status = JavaThreadStatus::NEW;
       break;
     case STARTED :
-    case RUNNABLE :
     case RUNNING :
     case PARKING :
     case TIMED_PARKING:
+    case UNPARKED:
     case YIELDING :
+    case YIELDED:
```
Comment (sspitsyn) [inline src/hotspot/share/classfile/javaClasses.cpp:1998]: Is this new state `YIELDED` really needed?
It looks like `UNPARKED` is also appeared where `YIELDED` is used.

File: src/java.base/share/classes/java/lang/VirtualThread.java (hunk 1 above)
Comment (AlanBateman) [reply src/java.base/share/classes/java/lang/VirtualThread.java:116]: > Just want to make sure this change is intentional. Before the comment was: `// runnable-mounted`.

Well spotted, the wrong comment line was moved.

File: src/hotspot/share/classfile/javaClasses.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlanBateman) [reply src/hotspot/share/classfile/javaClasses.cpp:1998]: I need to check the Hotspot style guide but I think the prevailing convention is no space between the label and the colon.

**Review by sspitsyn — APPROVED** [review]
The fix looks good to me.
Also, it can be more safe to run mach5 tier6 as well.

**Review by dholmes-ora — COMMENTED** [review]
I find the switch from `RUNNABLE` to `UNPARKED` somewhat unnecessary. I don't see any problem with `RUNNABLE` that needed fixing. I found the distinction and transition between `RUNNABLE` and `RUNNING` to be very clear. ??

**Comment by dholmes-ora** [comment]
Sorry I've deleted my earlier embarassingly stupid comment. The key point of this change is to split RUNNABLE into two states to indicated "runnable after yielding" and "runnable after being unparked".

**Pull request closed without merging**